.gitlab/includes/gcc12_cuda12_pipeline.yml

```diff
@@ -0,0 +1,76 @@
+# file BOOST_LICENSE_1_0.rst or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+include:
+  - local: '.gitlab/includes/common_pipeline.yml'
+  - local: '.gitlab/includes/common_spack_pipeline.yml'
+
+.variables_gcc12_cuda12_config:
+  variables:
+    ARCH: linux-ubuntu22.04-broadwell
+    BUILD_TYPE: Debug
+    COMPILER: gcc@12.1.0
+    CXXSTD: 17
+    SPACK_SPEC: "pika@main arch=$ARCH %${COMPILER} +cuda malloc=system cxxstd=$CXXSTD \
+                 ^boost@1.82.0 ^hwloc@2.9.1 ^cuda@12.0.0 ^fmt@10.0.0"
+
+gcc12_cuda12_spack_image:
+  stage: spack_configs
+  needs: [base_spack_image]
+  extends:
+    - .container-builder
+    - .variables_gcc12_cuda12_config
+  before_script:
+    - CONFIG_TAG=`echo $SPACK_SPEC | sha256sum - | head -c 16`
+    - compiler=${COMPILER/@/-}
+    - export PERSIST_IMAGE_NAME=$CSCS_REGISTRY_PATH/pika-$compiler:$CONFIG_TAG
+    - echo -e "compiler=$compiler\nBASE_IMAGE=$PERSIST_IMAGE_NAME" >> compiler.env
+  variables:
+    BASE_IMAGE: $CSCS_REGISTRY_PATH/pika-spack-base:$SPACK_VERSION
+    DOCKERFILE: .gitlab/docker/Dockerfile.spack_compiler
+    DOCKER_BUILD_ARGS: '["BASE_IMAGE","ARCH","COMPILER","SPACK_SPEC"]'
+  artifacts:
+    reports:
+      dotenv: compiler.env
+
+gcc12_cuda12_debug_build:
+  stage: build
+  extends:
+    - .container-builder
+    - .variables_gcc12_cuda12_config
+    - .cmake_variables_common
+  needs:
+    - gcc12_cuda12_spack_image
+  before_script:
+    - build_type=`echo $BUILD_TYPE | tr '[:upper:]' '[:lower:]'`
+    - configuration=$CMAKE_COMMON_FLAGS-$CMAKE_FLAGS-$CI_COMMIT_SHORT_SHA
+    - configuration=${configuration//-D/}
+    - CONFIG_TAG=`echo $configuration | sha256sum - | head -c 16`
+    - export PERSIST_IMAGE_NAME=$CSCS_REGISTRY_PATH/pika-$compiler-$build_type-build:$CONFIG_TAG
+    - echo -e "PERSIST_IMAGE_NAME=$PERSIST_IMAGE_NAME" > build.env
+  variables:
+    DOCKERFILE: .gitlab/docker/Dockerfile.spack_build
+    CMAKE_FLAGS: "-DCMAKE_BUILD_TYPE=$BUILD_TYPE -DPIKA_WITH_CXX_STANDARD=$CXXSTD \
+                  -DPIKA_WITH_CUDA=ON -DPIKA_WITH_MALLOC=system \
+                  -DCMAKE_CUDA_ARCHITECTURE=60 -DPIKA_WITH_CXX_LAMBDA_CAPTURE_DECLTYPE=OFF"
+# PIKA_WITH_CXX_LAMBDA_CAPTURE_DECLTYPE is OFF to test the fallback implementation of PIKA_FORWARD.
+    DOCKER_BUILD_ARGS: '["BASE_IMAGE","SOURCE_DIR","BUILD_DIR","CMAKE_COMMON_FLAGS","CMAKE_FLAGS"]'
+  artifacts:
+    reports:
+      dotenv: build.env
+
+gcc12_cuda12_debug_test:
+  extends:
+    - .variables_gcc12_cuda12_config
+    - .test_common_gpu
+  needs:
+    - gcc12_cuda12_debug_build
+  script:
+    - spack arch
+    - spack build-env $spack_spec -- bash -c "ctest --label-exclude COMPILE_ONLY --test-dir ${BUILD_DIR} -j$(nproc) --timeout 120 --output-on-failure --no-compress-output --no-tests=error"
+      # All async_cuda tests are disabled from running because the driver on Piz Daint is too old
+      # for CUDA 12
+  image: $PERSIST_IMAGE_NAME
```

Comment:
This test stage will not work on daint because CUDA 12 requires a newer driver which daint doesn't have (specifically, the CUDA tests won't run). We can leave it out for now, and consider running on clariden/hohgant later. The jenkins CI configuration with CUDA 12 also doesn't _run_ the CUDA tests.